lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -722,3 +729,62 @@ hex = convertToBase Base16
 -- | Decode a 'ByteString' from base16
 fromHex :: ByteArray bout => ByteString -> Either String bout
 fromHex = convertFromBase Base16
+
+data ErrUnXPrvStripPub
+    = ErrNoRoundtripMismatch
+      -- ^ The resulting bytestring would have been unable to roundtrip using
+      -- @xPrvFromStrippedPubXPrv@. Most likely because the input @XPrv@ was
+      -- encrypted.
+    | ErrUnexpectedRoundtripError
+      -- ^ Something unexpected went wrong.
```

Comment:
This second branch is quite dubious. There's nothing unexpected here, it feels more like an invariant violation. This branch is reached because of the check using `xPrvFromStrippedPubXPrv` if the given bytes are not of the right length. However, these bytes are produced by `unXPrvStripPub` itself, so it's more of a implementor error if we ever reach that point and shouldn't be pushed onto the function interface. 

So basically, `Left _ -> Left ErrUnexpectedRoundtripError` should actually really be:

```hs
    Left _ -> error "impossible: this state cannot be reached from a rightfully crafted XPrv". 
```

Or better, change the way the check is written to get rid of this impossible pattern match:

```hs
unXPrvStripPub k = do
    let res = stripPub . unXPrv $ k

    unless (Right res == roundtrip res) $ 
      Left ErrRoundtripFailed

    pure res
  where
    roundtrip = fmap unXPrv . xPrvFromStrippedPubXPrv
```

Reply:
If `unXPrvStripPub` were to only be used from the CLI, the extra `ErrUnexpectedRoundtripError` would do no damage, and provide _slightly_ neater/controllable output… But sure — invariant might be more sane.

>  Or better, change the way the check is written to get rid of this impossible pattern match:

_When_ the impossible happens, I'd rather see a `error "impossible` than have it disguised as a `ErrRoundtripFailed`.

Reply:
Also opting to use `case` instead of `unless` because I want to differentiate between `ErrRoundtripFailed` and `error "impossible`